Initial add of LCGP as submodule

mosesyhc opened the pull request:
This PR adds the LCGP package as a submodule.

For Moses:
- [x] Update BAND SDK

For reviewers:
- [x] Please install and test the package using instructions in [LCGP installation](https://lcgp.readthedocs.io/en/latest/#installation).  
- [x] Verify that `lcgp.__version__` returns `v0.2.0`.
- [x] _Optional test_. In addition, you may run the example script to test the package usage, shown below:

```
pip install matplotlib pandas
python ./illustration-examples/lcgp-3d.py
```
  
  This script should return a subdirectory `results_figures` with two plots, and a console output of error measures.  An example plot is shown here:
<img width="1000" height="500" alt="image" src="https://github.com/user-attachments/assets/b3f2c2f6-967d-463b-bf92-cb895429877f" />


**Note:** There are known runtime warnings in using tensorflow.

mosesyhc committed:
initial add of LCGP as submodule

mosesyhc committed:
updates lcgp citation and band sdk

beykyle reviewed (approved):
Successfully pip installed, ran tests, and ran the illustrative example.

wildsm reviewed (changes requested):
Started new venv

`pip install lcgp`

succeeded. Within python I then 

` >>> import lcgp`
 
which pointed out that pytest is a missing dependency:

`ModuleNotFoundError: No module named 'pytest'`

I addressed this dependency and continued. Note that I received the warning:

`.venv/lib/python3.12/site-packages/gpflow/versions.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.`


I then continued following the readme:

```
>>> lcgp.__version__
'0.2.0'
```



```
>>> lcgp.test()

...

=============================== warnings summary ===============================
.venv/lib/python3.12/site-packages/lcgp/tests/test_initialize.py::TestInit::test_simplest_1D_fail
  /home/wild/oldcomp/repos/bandframework/.venv/lib/python3.12/site-packages/tensorflow_probability/python/__init__.py:57: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if (distutils.version.LooseVersion(tf.__version__) <

.venv/lib/python3.12/site-packages/lcgp/tests/test_initialize.py::TestInit::test_simplest_1D_fail
  /home/wild/oldcomp/repos/bandframework/.venv/lib/python3.12/site-packages/tensorflow_probability/python/__init__.py:58: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    distutils.version.LooseVersion(required_tensorflow_version)):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 36 passed, 2 warnings in 9.39s ========================
True
```

I have created  PR https://github.com/mosesyhc/LCGP/pull/13 to address to minor issues in the readme.

I suggest addressing the missing pytest dependency and also being clear what you mean by "within the root directory". If someone just pip installed lcgp, where should they go?


My next review will do the optional tests.

wildsm reviewed (commented):
Now I needed to be within the lcgp codebase for the illustrative-examples directory to be in the expected place.

I got the usual warnings and error about my sandboxed system not having cuda (pasting just for awareness):

```
$ python ./illustration-examples/lcgp-3d.py
2025-09-16 10:15:14.763356: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2025-09-16 10:15:14.764026: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-09-16 10:15:14.858683: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2025-09-16 10:15:16.959883: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-09-16 10:15:16.960890: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
/home/wild/oldcomp/repos/bandframework/.venv/lib/python3.12/site-packages/gpflow/versions.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  import pkg_resources
2025-09-16 10:15:22.607691: E external/local_xla/xla/stream_executor/cuda/cuda_platform.cc:51] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)
```

This resulted in 
`RMSE: {200: 0.5304329570809118, 400: 0.4313014973742669}
`
and 2 figures in the promised location. Here is n=400:

<img width="1000" height="500" alt="lcgp-illustration-n-400" src="https://github.com/user-attachments/assets/cf10940d-c1a8-4cb3-8b06-12c5cf69d2c1" />

wildsm reviewed (changes requested):
Shouldn't there also be updates in this PR to:

- https://github.com/bandframework/bandframework/blob/develop/README.md : Include lcgp in the tools list.
- https://github.com/bandframework/bandframework/blob/develop/resources/sdkpolicies/README.md : A link to the lcgp's bandsdk

wildsm reviewed (changes requested):
This is now a review of the bandsdk at 
https://github.com/mosesyhc/LCGP/blob/9f1cc0e7a6fe1745f7cd7a0ac8ca8f62a3122b23/lcgp-bandsdk.md

I believe that you may want to change the first two links that are there. If people access your repo directly, the current relative links fail.

I would say the M2 is Partial. The readme is technically not in the top level directory but your explanation is great.

For R6, I suggest repeating the location of the readme (like you did in M2).

Everything else looks great!

mosesyhc committed:
include lcgp in band readme and sdkpolicies readme.

mosesyhc committed:
updating LCGP to v0.2.1

mosesyhc committed:
satisfy spellcheck for lcgp

mosesyhc committed:
update lcgp sdk link

mosesyhc commented:
Thank you both! 
@wildsm : 
- I have updated the `pytest` dependency in lcgp, and issued a new minor version. 
- Added lcgp to both BAND and SDK readme's.
- I fixed the relative links in the SDK.

Slightly puzzled by the [failed action](https://github.com/bandframework/bandframework/actions/runs/17776869205/job/50526653397#step:4:53).  The scan for links found the LCGP and resolved the reference but there is an error not resolving the link text?

wildsm committed:
Update bandframework-ci.yml

wildsm commented:
I fixed the CI note; there is an issue with the sdks and so they appear in blacklist.

wildsm commented:
Tests are passing and I've reviewed the latest changes. Merging now.
Thanks for the contribution @mosesyhc !

wildsm reviewed (approved):
Changes look great, thank you
